### DISCUSSION TITLE
Playing tone with function generator

### DISCUSSION THREAD

**Discussion by petersaj** [bonsai-rx/bonsai#1721] [General]
Hi - we're trying to play a sine tone using  `Function Generator`, but we're not sure exactly why we see certain things in our workflows. 

We've made some small examples, where we want to make a waveform for 1/4 second at frequency = 500, so our `Function Generator` parameters look like this: 

![image](https://github.com/bonsai-rx/bonsai/assets/17985056/7fde12d6-19df-4954-a88e-82f647a83bab)

If we just do `Function Generator` to `Audio Playback`, this plays a tone forever, though sometimes there's a skip in the tone near the beginning, and we don't understand why this plays forever, where the small skip comes from, and why it's sometimes there and sometimes not? 

![image](https://github.com/bonsai-rx/bonsai/assets/17985056/25b58d91-e001-4e65-8ea8-b07ae2afd72e)

If we try to pair the `Function Generator` with a `Take`, this only plays a tone if Take > 1 (Take = 1 shows the waveform, but no audio plays): 

![image](https://github.com/bonsai-rx/bonsai/assets/17985056/38673d22-32d6-4188-8e6c-c528d53d7890)

If we set `Take` = 2 and a `Repeat` at the end of that, the tone isn't continuous, so with the above it seems as if the first element in the `Take` the sequence is blank: 

![image](https://github.com/bonsai-rx/bonsai/assets/17985056/8765493d-d900-49b9-b3e9-59d09ba20dbf)

Another issue with the workflow above that we hear a 'click' at the offset of a tone.

Could you help us understand why the above things are happening, and what the best practice is for giving a consistent tone for a set amount of time or continuously?

**Comment by bruno-f-cruz** [comment]
Please add the workflow with the examples. thanks

**Comment by petersaj** [comment]
Sure - attached here (with ".txt" appended to upload), along with the questions for each:

- tone.bonsai: why does this play forever and smoothly?
[tone.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/14780412/tone.bonsai.txt)
- tone_take1.bonsai: why doesn't this play anything? 
[tone_take1.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/14780415/tone_take1.bonsai.txt)
- tone_take2.bonsai: why does this have a gap at the start and an audio artifact at the end of the tone?
[tone_take2.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/14780417/tone_take2.bonsai.txt)
- tone_take2_repeat.bonsai: why is this one non-continuous (and with an audio artifact at tone onset like above)?
[tone_take2_repeat.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/14780419/tone_take2_repeat.bonsai.txt)

**Reply by bruno-f-cruz** [reply]
This is a great question and a great exercise. I will try to give a complete answer but let me know if something is not clear.
First, one must understand that the FunctionGeneration outputs a function in smaller buffers that, when concatenated, make up the configured waveform. For instance, in your first example both Frequency and BufferSize are set to 44100. This essentially means that we are expecting one buffer object (OpenCV.Mat), with 441000 samples,  per second.

Now that we got that out of the way:

1. Function generator, when no input is provided, will attempt to generate its own data by honorting the SampleRate and BufferLength properties. In this case it will generate these at 1 hertz. The node will just keep generating the waveform and emitting the buffers forever and it will thus just play the sound endlessly.
2. The problem here stems from closing the observable sequence before the sink (AudioPlayback) has time to play your sound. Since Take emits an OnComplete as soon as the first buffer is propagated, I suspect that AudioCapture is being disposed before it even has the chance to playback anything. We can show this by preventing the OnComplete event from being propagated to the Sink using a BehaviorSubject:
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/d03f8423-5989-4553-99c3-0d8ce1f708d5)

3. Uhm not sure what do you mean. On my PC it plays the first buffer just fine and then it stops once the second one is emitted (see 2. for the explanation)

4. This could be one of two things. Clipping due to the buffers not concatenating properly (once you close the sequence and resubscribe the function generator resets its phase) OR the fact that you are closing and opening the connection to the audio card. I suspect its the latter since changing the order of the operators gets rid of the alias on my end:
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/e6680519-ec47-49fd-80a0-1c031de78c76)

Since you have a "Take", once again you close the audio playback and repeat resubscribes to the whole sequence again. You might have a gap corresponding to the time it takes to dispose and reinitialize the hardware.

In general, I think you want to use the pattern with the BehaviorSubject I described above. This way you can effectively isolate the signal generation from the hardware control without having to worry about disposing/initializing the connection multiple times in the same workflow.

**Comment by petersaj** [comment]
Great, very helpful explanations, thanks!
